src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,600 @@
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Linq;
+using Catalyst.Abstractions.Cryptography;
+using Catalyst.Core.Lib.Extensions;
+using Catalyst.Core.Modules.Cryptography.BulletProofs;
+using Catalyst.Protocol.Deltas;
+using Catalyst.Protocol.Transaction;
+using Google.Protobuf;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Core.Specs;
+using Nethermind.Dirichlet.Numerics;
+using Nethermind.Evm;
+using Nethermind.Evm.Precompiles;
+using Nethermind.Evm.Tracing;
+using Nethermind.Store;
+using Serilog;
+using Serilog.Events;
+
+namespace Catalyst.Core.Modules.Kvm
+{
+    /// <summary>
+    ///     Delta executor is responsible for executing delta entries code on the VM. Its responsibility is to prepare
+    ///     the execution environment and initial state of the VM as well as support some contract deployment operations.
+    ///     There is a minor responsibility issue in the way the code deposits are handled - depending on the execution
+    ///     level the deposit may happen inside the <see cref="VirtualMachine" /> or <see cref="DeltaExecutor" /> code.
+    /// </summary>
+    public class DeltaExecutor : IDeltaExecutor
+    {
+        private readonly ICryptoContext _cryptoContext = new FfiWrapper();
+        private readonly ILogger _logger;
+        private readonly ISpecProvider _specProvider;
+        private readonly IStateProvider _stateProvider;
+        private readonly IStorageProvider _storageProvider;
+        private readonly IVirtualMachine _virtualMachine;
+
+        /// <summary>
+        ///     Note that there is a distinct approach to state and storage even as only together they form the 'state'
+        ///     in the business sense. <see cref="IStorageProvider" /> needs to handle storage trees for various accounts
+        ///     while <see cref="IStateProvider" /> handles the basic accounts state with storage roots only.
+        /// </summary>
+        /// <param name="specProvider">The network upgrade spec - defines the virtual machine version.</param>
+        /// <param name="stateProvider">Access to accounts.</param>
+        /// <param name="storageProvider">Access to accounts' storage.</param>
+        /// <param name="virtualMachine">A virtual machine to execute the code on.</param>
+        /// <param name="logger">Logger for the execution details.</param>
+        public DeltaExecutor(ISpecProvider specProvider,
+            IStateProvider stateProvider,
+            IStorageProvider storageProvider,
+            IVirtualMachine virtualMachine,
```

Comment:
Should this be IKvm to match Autofac registration in KvmModule?

Reply:
fixed, committing in a moment

Reply:
You could inject IKVM to something that takes IVirtualMachine as IKVM extends IVirtualMachine but you would loose anything defined in IKVM